File: ruby/spec/lookup.spec.rb
input = <<~DOC.strip
color: cyan
close:up
# notes
-- long
is
-- long
DOC

scenarios = [
  {
    arguments: [3], # 'o'
    key: 'color',
    range: :key
  },
  {
    arguments: [0, 3], # 'o'
    key: 'color',
    range: :key
  },
  {
    arguments: [6], # ' '
    key: 'color',
    range: :element_operator
  },
  {
    arguments: [0, 6], # ' '
    key: 'color',
    range: :element_operator
  },
  {
    arguments: [7], # 'c'
    key: 'color',
    range: :value
  },
  {
    arguments: [0, 7], # 'c'
    key: 'color',
    range: :value
  },
  {
    arguments: [18], # 'u'
    key: 'close',
    range: :value
  },
  {
    arguments: [1, 6], # 'u'
    key: 'close',
    range: :value
  },
  {
    arguments: [21], # '#'
    key: 'notes',
    range: :section_operator
  },
  {
    arguments: [2, 0], # '#'
    key: 'notes',
    range: :section_operator
  },
  {
    arguments: [27], # 's'
    key: 'notes',
    range: :key
  },
  {
    arguments: [2, 6], # 's'
    key: 'notes',
    range: :key
  },
  {
    arguments: [37],
    key: 'long', # 'i'
    range: :value
  },
  {
    arguments: [4, 0],
    key: 'long', # 'i'
    range: :value
  },
  {
    arguments: [45],
    key: 'long', # 'n'
    range: :key
  },
  {
    arguments: [5, 5],
    key: 'long', # 'n'
    range: :key
  }
]

describe Enolib::Section do
  describe '#lookup' do
    scenarios.each do |scenario|
      context "at (#{scenario[:arguments].join(', ')})" do
        lookup =
          if scenario[:arguments].length > 1
            Enolib.lookup(input, line: scenario[:arguments][0], column: scenario[:arguments][1])
          else
            Enolib.lookup(input, index: scenario[:arguments][0])
          end

        it "looks up element '#{scenario[:element]}'" do
          expect(lookup[:element].string_key).to eq(scenario[:key])
        end

        it "looks up token '#{scenario[:zone]}'" do
          expect(lookup[:range]).to eq(scenario[:range])
        end
      end
    end
  end
end

File: php/spec/lookup/lookup.spec.php
<?php declare(strict_types=1);

use Eno\Parser;

const SCENARIOS = [
  [
    'arguments' => [3], # 'o'
    'element' => 'color',
    'zone' => 'key'
  ],
  [
    'arguments' => [0, 3], # 'o'
    'element' => 'color',
    'zone' => 'key'
  ],
  [
    'arguments' => [6], # ' '
    'element' => 'color',
    'zone' => 'element_operator'
  ],
  [
    'arguments' => [0, 6], # ' '
    'element' => 'color',
    'zone' => 'element_operator'
  ],
  [
    'arguments' => [7], # 'c'
    'element' => 'color',
    'zone' => 'value'
  ],
  [
    'arguments' => [0, 7], # 'c'
    'element' => 'color',
    'zone' => 'value'
  ],
  [
    'arguments' => [18], # 'u'
    'element' => 'close',
    'zone' => 'value'
  ],
  [
    'arguments' => [1, 6], # 'u'
    'element' => 'close',
    'zone' => 'value'
  ],
  [
    'arguments' => [21], # '#'
    'element' => 'notes',
    'zone' => 'section_operator'
  ],
  [
    'arguments' => [2, 0], # '#'
    'element' => 'notes',
    'zone' => 'section_operator'
  ],
  [
    'arguments' => [27], # 's'
    'element' => 'notes',
    'zone' => 'key'
  ],
  [
    'arguments' => [2, 6], # 's'
    'element' => 'notes',
    'zone' => 'key'
  ],
  [
    'arguments' => [37],
    'element' => 'long', # 'i'
    'zone' => 'content'
  ],
  [
    'arguments' => [4, 0],
    'element' => 'long', # 'i'
    'zone' => 'content'
  ],
  [
    'arguments' => [45],
    'element' => 'long', # 'n'
    'zone' => 'key'
  ],
  [
    'arguments' => [5, 5],
    'element' => 'long', # 'n'
    'zone' => 'key'
  ]
];

describe('Section', function() {
  describe('lookup()', function() {
    beforeAll(function() {
      $input = <<<DOC
color: cyan
close:up
# notes
-- long
is
-- long
DOC;

      $this->document = Parser::parse($input);
    });

    foreach(SCENARIOS as $scenario) {
      $position = join(', ', $scenario['arguments']);

      describe("at {$position}", function() use($scenario) {
        beforeAll(function() use($scenario) {
          $this->lookup = $this->document->lookup(...$scenario['arguments']);
        });

        it("looks up element '{$scenario['element']}'", function() use($scenario) {
          expect($this->lookup['element']->stringKey())->toEqual($scenario['element']);
        });

        it("looks up token '{$scenario['zone']}'", function() use($scenario) {
          expect($this->lookup['zone'])->toEqual($scenario['zone']);
        });
      });
    }
  });
});
